bitmind/validator/reward.py

```diff
@@ -4,69 +4,86 @@
 # Copyright © 2023 BitMind
 
 # Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
-# documentation files (the “Software”), to deal in the Software without restriction, including without limitation
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 # the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
 # and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 
 # The above copyright notice and this permission notice shall be included in all copies or substantial portions of
 # the Software.
 
-# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
 # THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 # THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 # OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 # DEALINGS IN THE SOFTWARE.
 
 
-from typing import List
+from typing import List, Dict, Tuple, Any
 import bittensor as bt
 import numpy as np
 
 
 def compute_penalty(y_pred: float) -> float:
+    """
+    Compute penalty for predictions outside valid range.
+
+    Args:
+        y_pred (float): Predicted value
+
+    Returns:
+        float: 0.0 if prediction is invalid, 1.0 if valid
+    """
     bad = (y_pred < 0.0) or (y_pred > 1.0)
     return 0.0 if bad else 1.0
 
 
 def get_rewards(
-        label: float,
-        responses: List[float],
-        uids: List[int],
-        axons: List[bt.axon],
-        challenge_modality: str,
-        performance_trackers,
-    ) -> np.array:
+    label: float,
+    responses: List[float],
+    uids: List[int],
+    axons: List[bt.axon],
+    challenge_modality: str,
+    performance_trackers: Dict[str, Any]
+) -> Tuple[np.ndarray, List[Dict[str, Dict[str, float]]]]:
```

Comment:
is this necessary to also return metrics in get_rewards?

Reply:
Yep get_rewards has been the function responsible for using the miner_performance_tracker object to compute metrics